File: problem/Min_Distance_Word_Fingers/minimumDistance.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

#define MIN(a, b) ((a) < (b) ? (a) : (b))

int cal(int a, int b)
{
    return abs(a / 6 - b / 6) + abs(a % 6 - b % 6);
}

int minimumDistance(char *s)
{
    int n = strlen(s);
    int dp[26] = {0};
    int ndp[26];

    for (int i = 1; i < n; i++)
    {
        int p = s[i - 1] - 'A';
        int t = s[i] - 'A';

        for (int j = 0; j < 26; j++)
        {
            ndp[j] = dp[j] + cal(p, t);
        }
        for (int j = 0; j < 26; j++)
        {
            ndp[p] = MIN(ndp[p], dp[j] + cal(j, t));
        }
        memcpy(dp, ndp, sizeof(dp));
    }

    int res = dp[0];
    for (int i = 1; i < 26; i++)
    {
        res = MIN(res, dp[i]);
    }
    return res;
}

File: problem/Min_Distance_Word_Fingers/minimumDistance.js
/**
 * @param {string} word
 * @return {number}
 */
var minimumDistance = function(word) {
    const cal = (a, b) => {
        return Math.abs(Math.floor(a / 6) - Math.floor(b / 6)) + Math.abs((a % 6) - (b % 6));
    };

    let n = word.length;
    let dp = new Array(26).fill(0);

    for (let i = 1; i < n; i++) {
        let ndp = new Array(26);
        let p = word.charCodeAt(i - 1) - 65;
        let t = word.charCodeAt(i) - 65;

        for (let j = 0; j < 26; j++) {
            ndp[j] = dp[j] + cal(p, t);
        }
        for (let j = 0; j < 26; j++) {
            ndp[p] = Math.min(ndp[p], dp[j] + cal(j, t));
        }
        dp = ndp;
    }

    return Math.min(...dp);
};